### PR TITLE
chore: use more explicit awaits when returning in client

### PR DESCRIPTION
**Pull request by mxschmitt**
This fixes "Problem 1" of https://github.com/microsoft/playwright/issues/28314.

### CODE REVIEW

**Review by dgozman — COMMENTED** [review]
I am not sure this is the right rule for us. We want to make sure all async function calls are awaited, not that we have at least one await in a function.

File: package.json [hunk 1]
```diff
@@ -85,7 +86,8 @@
     "electron-to-chromium": "^1.4.212",
     "enquirer": "^2.3.6",
     "esbuild": "^0.18.11",
-    "eslint": "^8.37.0",
+    "eslint": "^8.55.0",
+    "eslint-plugin-internal-playwright": "file:utils/eslint-plugin-internal-playwright",
```
Comment (mxschmitt) [inline package.json:90]: Note: According to my research there seems to be no other way of consuming an ESLint rule without treating it as a local package.

File: packages/playwright-core/src/client/network.ts [hunk 2]
```diff
@@ -159,16 +159,16 @@ export class Request extends ChannelOwner<channels.RequestChannel> implements ap
     return this._provisionalHeaders.headers();
   }
 
-  _actualHeaders(): Promise<RawHeaders> {
+  async _actualHeaders(): Promise<RawHeaders> {
     if (this._fallbackOverrides.headers)
-      return Promise.resolve(RawHeaders._fromHeadersObjectLossy(this._fallbackOverrides.headers));
+      return await Promise.resolve(RawHeaders._fromHeadersObjectLossy(this._fallbackOverrides.headers));
```
Comment (dgozman) [inline packages/playwright-core/src/client/network.ts:164]:
```suggestion
      return RawHeaders._fromHeadersObjectLossy(this._fallbackOverrides.headers);
```

File: packages/playwright-core/src/client/network.ts [hunk 3]
```diff
@@ -300,16 +300,16 @@ export class Route extends ChannelOwner<channels.RouteChannel> implements api.Ro
...
+  private async _raceWithTargetClose(promise: Promise<any>): Promise<void> {
     // When page closes or crashes, we catch any potential rejects from this Route.
     // Note that page could be missing when routing popup's initial request that
     // does not have a Page initialized just yet.
-    return this.request()._targetClosedScope().safeRace(promise);
+    return await this.request()._targetClosedScope().safeRace(promise);
   }
 
-  _startHandling(): Promise<boolean> {
+  async _startHandling(): Promise<boolean> {
```
Comment (dgozman) [inline packages/playwright-core/src/client/network.ts:310]: This one seems weird. I see we only use it in a single place, let's inline?

File: utils/eslint-plugin-internal-playwright/rules/await-promise-in-class-returns.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      'ClassDeclaration MethodDefinition ReturnStatement': function (statement) {
+        if (statement.type === 'ReturnStatement' && statement.argument) {
+          if (tsutils.isThenableType(
+            parserServices.program.getTypeChecker(),
+            statement.argument,
+            parserServices.getTypeAtLocation(statement.argument)
+          )) {
+            context.report({
+              node: statement,
+              message: 'Return statement in a class should await a promise',
```
Comment (dgozman) [inline utils/eslint-plugin-internal-playwright/rules/await-promise-in-class-returns.js:42]: Let's explain why to avoid confusion.

**Review by dgozman — APPROVED** [review]

File: packages/playwright-core/src/client/network.ts (hunk 3 above)
Comment (mxschmitt) [reply packages/playwright-core/src/client/network.ts:310]: I think its a layer violation otherwise. So far the promise wasn't exposed.

**Pull request merged**